bohr/snorkel_utils.py

```diff
@@ -1,20 +1,334 @@
+        return False
+
+    def match_bigram(self, stemmed_bigrams: Set[Tuple[str, str]]) -> bool:
+        for issue in self.__issues:
+            if not issue.stem_bigrams.isdisjoint(stemmed_bigrams): return True
+        return False
+
+    def contains_bigram(self, stemmed_bigram: Tuple[str, str]) -> bool:
+        for issue in self.__issues:
+            if stemmed_bigram in issue.stem_bigrams: return True
+        return False
+
+
+
+@dataclass
+class CommitFile:
+    filename: str
+    status: str
+    patch: Optional[str]
+    changes: Optional[str]
+
+
+class CommitFiles:
+    def __init__(self, files):
+        self.__files = files
+
+    def __len__(self) -> int:
+        return len(self.__files)
+
+    def __getitem__(self, idx) -> CommitFile:
+        return self.__files[idx]
+
+@dataclass
+class CommitMessage:
+    raw: str
+
+    @cached_property
+    def tokens(self) -> Set[str]:
+        if self.raw is None: return set()
+        return safe_word_tokenize(self.raw)
+
+    @cached_property
+    def ordered_stems(self) -> List[str]:
+        stemmer = PorterStemmer()
+        return [stemmer.stem(w) for w in self.tokens]
+
+    @cached_property
+    def stems(self) -> Set[str]:
+        return set(self.ordered_stems)
+
+    @cached_property
+    def stem_bigrams(self) -> Set[Tuple[str, str]]:
+        return set(bigrams(self.ordered_stems))
+
+    def match(self, stemmed_keywords: Set[str]) -> bool:
+        return not self.stems.isdisjoint(stemmed_keywords)
+
+    def match_bigram(self, stemmed_bigrams: Set[Tuple[str, str]]) -> bool:
+        return not self.stem_bigrams.isdisjoint(stemmed_bigrams)
+
+
+
+@dataclass
+class Commit:
+
+    owner: str
+    repository: str
+    sha: str
+    raw_message: str
+    message: CommitMessage = field(init=False)
+
+    def __post_init__(self):
+        self.message = CommitMessage(self.raw_message)
+
+    @lru_cache(maxsize=4)
+    def __load_df(self, type: str, owner: str, repository: str):
+        path = TRAIN_DIR / type  / owner / f"{repository}.csv"
+        if path.is_file():
+            return pd.read_csv(path, index_col=['sha'])
+        else:
+            return None
+
+    def __hash__(self):
+        return hash((self.owner, self.repository, self.sha))
+
+    @cached_property
+    def files(self) -> CommitFiles:
+        df = self.__load_df('files', self.owner, self.repository)
+        files = []
+
+        if df is not None:
+            try:
+                df = df.loc[[self.sha]]
+                for sha, file in df.iterrows():
+                    files.append(CommitFile(file.filename, file.status, file.get('patch', None), file.get('change', None)))
+            except KeyError as e:
+                pass
+
+        return CommitFiles(files)
+
+    @cached_property
+    def issues(self) -> Issues:
+        df = self.__load_df('issues', self.owner, self.repository)
+        issues = []
+
+        if df is not None:
+            df = df.loc[[self.sha]]
```

Comment:
crashes with `KeyError` if there is no issue that corresponds to the given commit